### PR TITLE
Add sanity checks for propensity score values

### PR DESCRIPTION
**Pull request by t-tte**
Addresses #107

### CODE REVIEW

**Comment by jeongyoonlee** [comment]
We can ignore failing tests due to missing imports from the previous merge. I fixed it in #108.

**Comment by ppstacy** [comment]
@t-tte thanks for adding the check here! I have a question regarding at least two distinct values for the propensity score. Recently when we use X/R-learner for the Bidder project, we used constant value (0.5) because of the equal chance for intervention in both the control and treatment group. So it probably will fail here.  Any suggestions for how should we handle this? 

cc @jeongyoonlee  who probably has more context here as well.

**Comment by t-tte** [comment]
Hi @ppstacy - that's a fair comment given that using a constant instead of a propensity score is suggested as an option in Kunzel et al (2019). That said, I think it would make sense to keep the propensity scores (and the associated checks) as one thing, and the option to pass a constant to the X-Learner as another thing. For example, perhaps we can add an optional argument to the X-Learner that lets the user fit the model with a constant instead of a propensity score.

File: causalml/inference/meta/utils.py [hunk 1]
```diff
@@ -25,6 +25,17 @@ def check_p_conditions(p, t_groups):
...
             'If p is passed as an np.ndarray, there must be only 1 unique non-control group in the treatment vector.'
+        assert len(np.unique(p)) > 2, \
+            'p should have more than two distinct values'
+        assert (0 < p).all() and (p < 1).all(), \
+            'The values of p should lie within the (0, 1) interval.'
+            
+    if isinstance(p, dict):
+        for t_name in t_groups:
+            assert len(np.unique(p[t_name])) > 2, \
+                'p should have more than two distinct values'
```
Comment (jeongyoonlee) [inline causalml/inference/meta/utils.py]: As @ppstacy pointed out, we may use a constant propensity e.g. in RCT. I don't think we need this check even for R-learner as long as propensity values are not extreme (i.e. close to either 0 or 1).

File: causalml/inference/meta/utils.py [hunk 2]
```diff
@@ -25,6 +25,17 @@ def check_p_conditions(p, t_groups):
     if isinstance(p, (np.ndarray, pd.Series)):
         assert t_groups.shape[0] == 1, \
             'If p is passed as an np.ndarray, there must be only 1 unique non-control group in the treatment vector.'
+        assert len(np.unique(p)) > 2, \
+            'p should have more than two distinct values'
+        assert (0 < p).all() and (p < 1).all(), \
+            'The values of p should lie within the (0, 1) interval.'
```
Comment (jeongyoonlee) [inline causalml/inference/meta/utils.py]: Should we consider using `0 + eps` and `1 - eps` for stability?

File: causalml/inference/meta/utils.py [hunk 3]
```diff
@@ -25,6 +25,17 @@ def check_p_conditions(p, t_groups):
...
+            
+    if isinstance(p, dict):
+        for t_name in t_groups:
+            assert len(np.unique(p[t_name])) > 2, \
+                'p should have more than two distinct values'
+            assert (0 < p[t_name]).all() and (p[t_name] < 1).all(), \
+                'The values of p should lie within the (0, 1) interval.'
```
Comment (jeongyoonlee) [inline causalml/inference/meta/utils.py]: Same as in line 30. Let's use `0 + eps` and `1 - eps` instead.

**Comment by peterfoley** [comment]
As a user of many tools that have tried to protect me from myself, I don’t like requiring >2 unique values. If for whatever reason a user has a propensity score with only 2 unique values, shouldn’t they still be allowed to use the package?

If folks are passing categorical values by mistake, they’ll probably have integer values, and the second assertion will catch those cases. Since propensity scores in most causal inference applications have to be in (0,1), that assertion matches user expectations.

File: causalml/inference/meta/utils.py [hunk 4]
```diff
@@ -20,11 +20,19 @@ def check_treatment_vector(treatment, control_name=None):
 
 
 def check_p_conditions(p, t_groups):
+    eps = np.finfo(float).eps
     assert isinstance(p, (np.ndarray, pd.Series, dict)), \
         'p must be an np.ndarray, pd.Series, or dict type'
     if isinstance(p, (np.ndarray, pd.Series)):
         assert t_groups.shape[0] == 1, \
             'If p is passed as an np.ndarray, there must be only 1 unique non-control group in the treatment vector.'
+        assert (0 + eps < p).all() and (p < 1 + eps).all(), \
```
Comment (jeongyoonlee) [inline causalml/inference/meta/utils.py]: `p < 1 - eps` instead of `p < 1 + eps`.

File: causalml/inference/meta/utils.py [hunk 5]
```diff
@@ -20,11 +20,19 @@ def check_treatment_vector(treatment, control_name=None):
...
+            'The values of p should lie within the (0, 1) interval.'
+
+    if isinstance(p, dict):
+        for t_name in t_groups:
+            assert (0 + eps < p[t_name]).all() and (p[t_name] < 1 + eps).all(), \
```
Comment (jeongyoonlee) [inline causalml/inference/meta/utils.py]: Same as above.

**Comment by jeongyoonlee** [comment]
LGTM. Thanks for fixing this over the weekend. Please try not to work too hard. :p

**Pull request merged**